param_sweep.py

```diff
@@ -0,0 +1,374 @@
+        plt.title(f'Security with parameters {desc_params}')
+
+        Logging.log('sweep', log_level, 'Saved the line plot graph to: %s',
+                    file_name + '_plot.png')
+        fig.savefig(file_name + '_plot.png')
+
+    elif len(axis_vars) == 2:
+        axis_vars = sorted(axis_vars.items(), key=lambda x: params[x[0]][1])
+        x_param = axis_vars[1][0]
+        x = sorted(params[x_param][0])
+        y_param = axis_vars[0][0]
+        y = sorted(params[y_param][0])
+
+        values = [i[1] for i in sorted(result_dict.items(), key=lambda x: x[0])]
+        mat = np.flip(np.array(values).reshape(
+            (len(set(y)), len(set(x)))),
+                      axis=0)
+
+        fig, ax = plt.subplots(figsize=(20, 20), dpi=80)
+
+        ax.imshow(mat)
+        ax.set_xticks(np.arange(0, len(set(x)), 1))
+        ax.set_yticks(np.arange(0, len(set(y)), 1))
+        ax.set_xticklabels(sorted(list(set(x))))
+        ax.set_yticklabels(sorted(list(set(y)), reverse=True))
+
+        desc_params = {i:v[0] for i, (v, _) in params.items() if len(v) == 1}
+        plt.title(f'Security with parameters {desc_params}')
+
+        for (j, i), label in np.ndenumerate(mat):
+            ax.text(i, j, round(label, 1), ha='center', va='center', color='white')
+
+        ax.set_xlabel(f'Parameter: {x_param}')
+        ax.set_ylabel(f'Parameter: {y_param}')
+
+        fig.savefig(file_name + '_gradient.png')
+        Logging.log('sweep', log_level, 'Saved the gradient graph to: %s',
+                    file_name + '_gradient.png')
+
+        if security_cutoff:
+            fig2, ax2 = plt.subplots(figsize=(20, 20), dpi=80)
+            binary_mat = mat >= security_cutoff
+            ax2.imshow(binary_mat)
+            ax2.set_xticks(np.arange(0, len(set(x)), 1))
+            ax2.set_yticks(np.arange(0, len(set(y)), 1))
+            ax2.set_xticklabels(sorted(list(set(x))))
+            ax2.set_yticklabels(sorted(list(set(y)), reverse=True))
+
+            for (j, i), label in np.ndenumerate(mat):
+                ax2.text(i,
+                         j,
+                         round(label, 1),
+                         ha='center',
+                         va='center',
+                         color='white' if label < security_cutoff else 'black')
+
+            for (j, i), label in np.ndenumerate(mat):
+                ax.text(i, j, round(label, 1), ha='center', va='center', color='white')
+            plt.title(f'Security with parameters {desc_params}'
+                      f' and cutoff at {security_cutoff}')
+
+            ax2.set_xlabel(f'Parameter: {x_param}')
+            ax2.set_ylabel(f'Parameter: {y_param}')
+
+            fig2.savefig(file_name + '_cutoff.png')
+            Logging.log('sweep', log_level, 'Saved the cutoff graph to: %s',
+                        file_name + '_cutoff.png')
+
+
+def PerformCalculation(input_params: tuple[int, float],
```

Comment:
This should be `perform_calculation`

Reply:
Maybe `estimatef`  or so? To make it clearer what this function does?

Reply:
Done - I combined this with the `security_level` function and just kept the `security_level` name.